### PR TITLE
Update Cookie.php

### PR DESCRIPTION
**Pull request by sarciszewski**
Don't compare hashes with ===, use a consant-time comparison function

### CODE REVIEW

File: classes/Kohana/Cookie.php [hunk 1]
```diff
@@ -83,6 +84,20 @@ public static function get($key, $default = NULL)
 
 		return $default;
 	}
+	
+	/**
+	 * Compare two hashes in a time-invariant manner.
+	 * Prevents cryptographic side-channel attacks (timing attacks, specifically)
+	 */
+	public static function compare($a, $b)
```
Comment (shadowhand) [inline classes/Kohana/Cookie.php:92]: this is awesome! can you make it more general purpose by putting it in `Security::slow_equals`?

File: classes/Kohana/Cookie.php [hunk 2]
```diff
@@ -83,6 +84,20 @@ public static function get($key, $default = NULL)
 
 		return $default;
 	}
+	
+	/**
+	 * Compare two hashes in a time-invariant manner.
+	 * Prevents cryptographic side-channel attacks (timing attacks, specifically)
```
Comment (shadowhand) [inline classes/Kohana/Cookie.php:90]: docblock is missing `@param` and `@return`

File: classes/Kohana/Cookie.php [hunk 3]
```diff
@@ -83,6 +84,20 @@ public static function get($key, $default = NULL)
...
+	 */
+	public static function compare($a, $b) 
+	{
+		$diff = strlen($a) ^ strlen($b);
+		for($i = 0; $i < strlen($a) && $i < strlen($b); $i++)
```
Comment (shadowhand) [inline classes/Kohana/Cookie.php:95]: we use `AND` not `&&` in Kohana

File: classes/Kohana/Cookie.php (hunk 1 above)
Comment (sarciszewski) [inline classes/Kohana/Cookie.php:92]: Uh, sure.

File: classes/Kohana/Cookie.php (hunk 2 above)
Comment (sarciszewski) [inline classes/Kohana/Cookie.php:90]: Sorry, I'm using the github editor and I'm spoiled by Netbeans auto-filling that in :P

**Comment by shadowhand** [comment]
This is a fantastic pull request, but it is targeting the wrong branch. Please reopen against `3.3/develop` and fix the issues noted. Thank you!

**Comment by cyrrill** [comment]
```
/**
 * Compares two strings $a and $b in length-constant time.
 *
 * Password Hashing With PBKDF2 (
```

http://crackstation.net/hashing-security.htm).
     \* Copyright (c) 2013, Taylor Hornby
     *
     \* @param string $a
     \* @param string $b
     \* @return bool
     */
    public static function slow_equals($a, $b)
    {
        $diff = strlen( $a ) ^ strlen( $b );
        for( $i = 0; $i < strlen( $a ) && $i < strlen( $b ); $i++ )
        {
            $diff |= ord( $a[$i] ) ^ ord( $b[$i] );
        }
        return $diff === 0;
    }

**Pull request closed without merging**